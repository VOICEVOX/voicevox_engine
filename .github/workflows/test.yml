name: test

on:
  push:
  pull_request:
  workflow_dispatch:

defaults:
  run:
    shell: bash

jobs:
  test:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-20.04, macos-latest, windows-latest]

    steps:
      - name: <Setup> Check out the repository
        uses: actions/checkout@v4

      - name: <Setup> Prepare Python Runtime / Python Dependencies
        uses: ./.github/actions/prepare_python
        with:
          requirements-suffix: "-dev"

      - name: <Test> Validate poetry.lock
        run: |
          poetry lock --no-update
          git diff --exit-code

      - name: <Test> Check dependency lists
        run: |
          poetry export --without-hashes -o requirements.txt.check
          poetry export --without-hashes --with dev -o requirements-dev.txt.check
          poetry export --without-hashes --with build -o requirements-build.txt.check

          diff -q requirements.txt requirements.txt.check || \
          diff -q requirements-dev.txt requirements-dev.txt.check || \
          diff -q requirements-build.txt requirements-build.txt.check > /dev/null
          if [ $? = 1 ]; then
            echo "poetry export has some diff"
            exit 1
          fi

      - name: <Test> Check format
        run: pysen run lint

      - name: <Test> Test codes and coverage
        run: coverage run --omit=test/* -m pytest

      - name: <Deploy> Submit coverage results to Coveralls
        if: matrix.os == 'ubuntu-20.04'
        run: coveralls --service=github
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

      - name: <Test> Check licenses
        run: OUTPUT_LICENSE_JSON_PATH=/dev/null bash tools/create_venv_and_generate_licenses.bash

      - name: <Test> Test names by checking typo
        if: matrix.os == 'ubuntu-20.04'
        uses: crate-ci/typos@v1.21.0

  lint-builders:
    runs-on: ubuntu-20.04
    steps:
      - name: <Setup> Check out the repository
        uses: actions/checkout@v4

      - name: <Setup> Install ShellCheck
        run: |
          sudo apt-get update
          sudo apt-get install -y shellcheck

      - name: <Test> Check shell files
        run: git ls-files | grep -E '\.(ba)?sh' | xargs shellcheck

      - name: <Test> Check workflow files
        run: |
          bash <(curl https://raw.githubusercontent.com/rhysd/actionlint/main/scripts/download-actionlint.bash)
          ./actionlint
