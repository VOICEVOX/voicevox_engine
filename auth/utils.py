from passlib.context import CryptContext
from datetime import datetime, timedelta
from typing import Union, Any
from jose import jwt
from auth.setting import (
    ACCESS_TOKEN_EXPIRE_MINUTES,
    JWT_SECRET_KEY,
    ALGORITHM,
    REFRESH_TOKEN_EXPIRE_MINUTES,
    JWT_REFRESH_SECRET_KEY
)

password_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

def get_hashed_password(password: str) -> str:
    return password_context.hash(password)

def verify_password(password: str, hashed_pass: str) -> bool:
    return password_context.verify(password, hashed_pass)

def create_access_token(subject: Union[str, Any], expires_delta: int = None) -> str:
    if expires_delta is not None:
        expires_delta = datetime.utcnow() + expires_delta
    else:
        expires_delta = datetime.utcnow() + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    
    to_encode = {"exp": expires_delta, "sub": str(subject)}
    encoded_jwt = jwt.encode(to_encode, JWT_SECRET_KEY, ALGORITHM)
    return encoded_jwt

def create_refresh_token(subject: Union[str, Any], expires_delta: int = None) -> str:
    if expires_delta is not None:
        expires_delta = datetime.utcnow() + expires_delta
    else:
        expires_delta = datetime.utcnow() + timedelta(minutes=REFRESH_TOKEN_EXPIRE_MINUTES)
    
    to_encode = {"exp": expires_delta, "sub": str(subject)}
    encoded_jwt = jwt.encode(to_encode, JWT_REFRESH_SECRET_KEY, ALGORITHM)
    return encoded_jwt

def payload_token(token: str):
    payload = jwt.decode(token, JWT_SECRET_KEY, algorithms=[ALGORITHM])
    return payload