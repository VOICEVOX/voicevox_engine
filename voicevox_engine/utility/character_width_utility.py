"""文字列の全角・半角に関する utility"""

import re
from typing import NewType, TypeGuard

# 半角アルファベット文字列を示す型
HankakuAlphabet = NewType("HankakuAlphabet", str)


def is_hankaku_alphabet(text: str) -> TypeGuard[HankakuAlphabet]:
    """文字列が半角アルファベットのみで構成されているかを判定する"""
    return bool(re.fullmatch("[a-zA-Z]+", text))


def convert_zenkaku_alphabet_to_hankaku(text: str) -> str:
    """全角アルファベットを半角に変換する"""
    return text.translate(
        str.maketrans(
            "".join(chr(0xFF01 + i) for i in range(94)),
            "".join(chr(0x21 + i) for i in range(94)),
        )
    )
