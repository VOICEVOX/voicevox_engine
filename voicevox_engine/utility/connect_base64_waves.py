import base64
import io
from typing import List, Tuple

import numpy as np
import soundfile
from scipy.signal import resample


class ConnectBase64WavesException(Exception):
    def __init__(self, message: str):
        self.message = message


def decode_base64_waves(waves: List[str]) -> List[Tuple[np.ndarray, int]]:
    """
    base64エンコードされた複数のwavデータをデコードする
    Parameters
    ----------
    waves: list[str]
        base64エンコードされたwavデータのリスト
    Returns
    -------
    waves_nparray_sr: List[Tuple[np.ndarray, int]]
        (NumPy配列の音声波形データ, サンプリングレート) 形式のタプルのリスト
    """
    if len(waves) == 0:
        raise ConnectBase64WavesException("wavファイルが含まれていません")

    waves_nparray_sr = []
    for wave in waves:
        try:
            wav_bin = base64.standard_b64decode(wave)
        except ValueError:
            raise ConnectBase64WavesException("base64デコードに失敗しました")
        try:
            _data = soundfile.read(io.BytesIO(wav_bin))
        except Exception:
            raise ConnectBase64WavesException("wavファイルを読み込めませんでした")
        waves_nparray_sr.append(_data)

    return waves_nparray_sr


def connect_base64_waves(waves: List[str]) -> Tuple[np.ndarray, int]:
    waves_nparray_sr = decode_base64_waves(waves)

    max_sampling_rate = max([sr for _, sr in waves_nparray_sr])
    max_channels = max([x.ndim for x, _ in waves_nparray_sr])
    assert 0 < max_channels <= 2

    waves_nparray_list = []
    for nparray, sr in waves_nparray_sr:
        if sr != max_sampling_rate:
            nparray = resample(nparray, max_sampling_rate * len(nparray) // sr)
        if nparray.ndim < max_channels:
            nparray = np.array([nparray, nparray]).T
        waves_nparray_list.append(nparray)

    return np.concatenate(waves_nparray_list), max_sampling_rate
