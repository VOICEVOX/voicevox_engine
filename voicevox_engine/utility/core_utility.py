import os


def get_half_logical_cores() -> int:
    logical_cores = os.cpu_count()
    if logical_cores is None:
        return 0
    return logical_cores // 2
