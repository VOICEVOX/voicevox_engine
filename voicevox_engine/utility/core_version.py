from typing import Iterable

from packaging.version import Version
from packaging.version import parse as parse_version


def parse_core_version(version: str) -> Version:
    return parse_version(version)


def get_latest_core_version(versions: Iterable[str]) -> str:
    if len(versions) == 0:
        raise Exception('versions must be non-empty.')

    return str(
        max(map(parse_core_version, versions))
    )
