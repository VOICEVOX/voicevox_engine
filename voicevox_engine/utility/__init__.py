from .connect_base64_waves import (
    ConnectBase64WavesException,
    connect_base64_waves,
    decode_base64_waves,
)
from .core_version_utility import get_latest_core_version, parse_core_version
from .mutex_utility import mutex_wrapper
from .path_utility import delete_file, engine_root, get_save_dir

__all__ = [
    "ConnectBase64WavesException",
    "connect_base64_waves",
    "decode_base64_waves",
    "get_latest_core_version",
    "parse_core_version",
    "delete_file",
    "engine_root",
    "get_save_dir",
    "mutex_wrapper",
]
