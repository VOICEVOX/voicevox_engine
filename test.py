import argparse
import wave
import numpy as np
from pathlib import Path
from voicevox_engine.core.core_initializer import initialize_cores
from voicevox_engine.tts_pipeline.tts_engine import make_tts_engines_from_cores, LATEST_VERSION
from voicevox_engine.utility.path_utility import engine_root, get_save_dir
from voicevox_engine.model import AudioQuery


# === CẤU HÌNH TRỰC TIẾP ===
HOST = ""
PORT = 0  # Không sử dụng API
USE_GPU = False
VOICEVOX_DIR          = Path("/home/trungquang/open_avt_phuc/OpenAvatarChat/src/handlers/tts/voicevox/voicevox_engine/voicevox_core")           # nếu có
VOICELIB_DIRS         = None
RUNTIME_DIRS          = [Path("/home/trungquang/open_avt_phuc/OpenAvatarChat/src/handlers/tts/voicevox/voicevox_engine/onnxruntime/onnxruntime-linux-x64-1.13.1/lib"),]
ENABLE_MOCK = False
INIT_PROCESSES = 2
LOAD_ALL_MODELS = True
CPU_NUM_THREADS = 4
DISABLE_MUTABLE_API = False
# ==========================



def synthesize_to_file(text: str, speaker: int, sample_rate: int, output_wav: Path):
    # Khởi tạo core và engine
    core_mgr = initialize_cores(
        use_gpu=USE_GPU,
        voicelib_dirs=VOICELIB_DIRS,
        voicevox_dir=VOICEVOX_DIR,
        runtime_dirs=RUNTIME_DIRS,
        cpu_num_threads=CPU_NUM_THREADS,
        enable_mock=ENABLE_MOCK,
        load_all_models=True,
    )
    tts_manager = make_tts_engines_from_cores(core_mgr)

    # Lấy engine phiên bản mới nhất
    engine = tts_manager.get_tts_engine(LATEST_VERSION)

    # Tạo query cho synthesis
    accent_phrases = engine.create_accent_phrases(text, speaker)
    query = AudioQuery(
        accent_phrases=accent_phrases,
        speedScale=1.0,
        pitchScale=0.0,
        intonationScale=1.0,
        volumeScale=1.0,
        prePhonemeLength=0.1,
        postPhonemeLength=0.1,
        pauseLength=None,
        pauseLengthScale=1.0,
        outputSamplingRate=sample_rate,
        outputStereo=False,
        kana="",
    )

    # Tổng hợp
    wav_array = engine.synthesize(query)
    # wav_array: numpy array shape (channels, frames)

    # Ghi file WAV
    samples = (wav_array * 32767).astype(np.int16)
    with wave.open(str(output_wav), 'wb') as wf:
        wf.setnchannels(samples.shape[0])
        wf.setsampwidth(2)
        wf.setframerate(sample_rate)
        wf.writeframes(samples.T.tobytes())


def main():
    parser = argparse.ArgumentParser(description="Một script đơn giản để tổng hợp giọng nói bằng VOICEVOX mà không cần API.")
    parser.add_argument('--text', type=str, help='Văn bản cần tổng hợp', default="不動産は")
    parser.add_argument('--speaker', type=int, help='ID của speaker', default=1)
    parser.add_argument('--sample_rate', type=int, help='Tỷ lệ mẫu', default=24000)
    parser.add_argument('--output', type=Path, help='Đường dẫn file WAV đầu ra', default="/home/trungquang/open_avt_phuc/OpenAvatarChat/src/handlers/tts/voicevox/output.wav")
    args = parser.parse_args()

    synthesize_to_file(
        text=args.text,
        speaker=args.speaker,
        sample_rate=args.sample_rate,
        output_wav=args.output,
    )
    print(f"Đã sinh file: {args.output}")


if __name__ == '__main__':
    main()
