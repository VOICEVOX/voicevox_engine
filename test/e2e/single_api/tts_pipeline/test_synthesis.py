"""
/synthesis API のテスト
"""

from test.e2e.single_api.utils import gen_mora
from test.utility import hash_wave_floats_from_wav_bytes

from fastapi.testclient import TestClient
from syrupy.assertion import SnapshotAssertion


def test_post_synthesis_200(client: TestClient, snapshot: SnapshotAssertion) -> None:
    query = {
        "accent_phrases": [
            {
                "moras": [
                    gen_mora("テ", "t", 2.3, "e", 0.8, 3.3),
                    gen_mora("ス", "s", 2.1, "U", 0.3, 0.0),
                    gen_mora("ト", "t", 2.3, "o", 1.8, 4.1),
                ],
                "accent": 1,
                "pause_mora": None,
                "is_interrogative": False,
            }
        ],
        "speedScale": 1.0,
        "pitchScale": 1.0,
        "intonationScale": 1.0,
        "volumeScale": 1.0,
        "prePhonemeLength": 0.1,
        "postPhonemeLength": 0.1,
        "pauseLength": None,
        "pauseLengthScale": 1.0,
        "outputSamplingRate": 24000,
        "outputStereo": False,
        "kana": "テ'_スト",
    }
    response = client.post("/synthesis", params={"speaker": 0}, json=query)
    assert response.status_code == 200

    # 音声波形が一致する
    assert response.headers["content-type"] == "audio/wav"
    assert snapshot == hash_wave_floats_from_wav_bytes(response.read())


def test_post_synthesis_old_audio_query_200(
    client: TestClient, snapshot: SnapshotAssertion
) -> None:
    """古いバージョンの audio_query でもエラーなく合成できる"""
    query = {
        "accent_phrases": [
            {
                "moras": [
                    gen_mora("テ", "t", 2.3, "e", 0.8, 3.3),
                    gen_mora("ス", "s", 2.1, "U", 0.3, 0.0),
                    gen_mora("ト", "t", 2.3, "o", 1.8, 4.1),
                ],
                "accent": 1,
                "pause_mora": None,
                "is_interrogative": False,
            }
        ],
        "speedScale": 1.0,
        "pitchScale": 1.0,
        "intonationScale": 1.0,
        "volumeScale": 1.0,
        "prePhonemeLength": 0.1,
        "postPhonemeLength": 0.1,
        "outputSamplingRate": 24000,
        "outputStereo": False,
    }
    response = client.post("/synthesis", params={"speaker": 0}, json=query)
    assert response.status_code == 200

    # 音声波形が一致する
    assert response.headers["content-type"] == "audio/wav"
    assert snapshot == hash_wave_floats_from_wav_bytes(response.read())
