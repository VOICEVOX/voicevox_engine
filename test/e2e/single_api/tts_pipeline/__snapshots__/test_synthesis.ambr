# serializer version: 1
# name: test_post_synthesis_200
  'MD5:f7d42ce5787856549abc3d2d7561c06f'
# ---
# name: test_post_synthesis_old_audio_query_200
  'MD5:f7d42ce5787856549abc3d2d7561c06f'
# ---
